lib/result.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
```

Comment:
Is this borrowed from another Shopify project?

Reply:
Yes, [Shopify ADT](https://github.com/Shopify/shopify-adt-gem), but that utilizes sorbet.

Reply:
I might keep the names the same as the original Shopify ADT. Since this functions so similarly to it I worry developers coming into use this Result would be thrown off by the method name changes.

Reply:
Hmm ... I see your point. My main concern is that `map` has a preconceived notion in Ruby – transforming a list of things into a list of other things. I will think about this a bit more.

Reply:
I'm thinking of removing `map` and `map_error` altogether. Instead, we could have a simple function for wrapping a lambda statement and auto constructing the result:

```ruby
result.and_then(Result.ok { |value| value.upcase! })
```

Alternatively, `and_then` could auto-construct a result altogether based on whether an exception was raised or a result was provided

```ruby
result.and_then { |value| value.length > 6 ? value : raise "Password too short" }
# equivalent to
result.and_then { |value| value.length > 6 ? Result.ok(value) : Result.error("Password too short") }
```

`and_then` already raises if you don't provide a result and therefore forces you to be deliberate about returning `Result.ok` or `Result.error`, but at the same time, we provide methods to circumvent having to be deliberate by providing `map` and `map_error`. Which when raising exceptions would break the entire chain rather than wrapping the exception and propagating it to the `or_else` handler. 😕

Reply:
I think `map` has value if you want to apply an operation that does not have a failure state. If I am adding 1 to an Integer I don't think I need to handle an error state for example.

I think `map` is actually a good name because it is the standard term in functional program. You are applying a function inside a container. In this case you are apply the function to the value held by Result and in the case you are describing you are applying it to the values held in Array. 

You can also see: https://en.wikipedia.org/wiki/Map_(higher-order_function)#:~:text=In%20many%20programming%20languages%2C%20map,when%20considered%20in%20functional%20form.

Reply:
I introduced `map` as discussed. 🙂